### PR TITLE
Fix pip idempotence in check mode

### PR DESCRIPTION
**Pull request by kustodian**
##### SUMMARY
PIP package names must be case insensitive, and must consider hyphens and underscores to be equivalent (https://www.python.org/dev/peps/pep-0426/#name), because of this the module didn't work correctly in check mode. For example if the passed package name had a different case or an underscore instead of a hyphen (or the other way around) compared to the installed package, check mode reported as changed, even though packages were installed. Now the module ignores case and hyphens/underscores in package names, so check mode works correctly.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
pip

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0
```

### CODE REVIEW

File: lib/ansible/modules/packaging/language/pip.py [hunk 1]
```diff
@@ -302,6 +302,15 @@ def _recover_package_name(names):
     return package_names
 
 
+def _normalize_package_name(name):
```
Comment (robinro) [inline lib/ansible/modules/packaging/language/pip.py]: This function does not behave the some way pip does. It would be better to use the some underlying code, see 
https://packaging.pypa.io/en/latest/_modules/packaging/utils/#canonicalize_name

It would be best to import this from some core library, but as far as I see it's not exposed, so for me it would also be ok to copy the code here with a reference to the source.

**Review by robinro — CHANGES_REQUESTED** [review]
@kustodian Thanks for this PR and the test!
This looks good and with the normalization function adapted it's ready to be included.

File: lib/ansible/modules/packaging/language/pip.py (hunk 1 above)
Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py]: Nice find will add it shortly. Thx!

**Comment by kustodian** [comment]
@robinro I made the changes you suggested and when the tests pass this should be ready for merging.

**Comment by robinro** [comment]
Thanks @kustodian 
shipit

**Comment by kustodian** [comment]
@Lujeni @mattupstate could you please give your approval (if you approve the change of course) so that we can get this merged?

**Comment by maxamillion** [comment]
rebuild_merge

**Comment by webknjaz** [comment]
cc @HD650

File: lib/ansible/modules/packaging/language/pip.py [hunk 2]
```diff
@@ -674,7 +682,8 @@ def main():
 
             changed = False
             if name:
-                pkg_list = [p for p in out.split('\n') if not p.startswith('You are using') and not p.startswith('You should consider') and p]
+                pkg_list = [p for p in out.split('\n')
```
Comment (webknjaz) [inline lib/ansible/modules/packaging/language/pip.py]: It looks like this hasn't been changed. Please don't do any unrelated changes. It's makes it hard do review PRs.

File: lib/ansible/modules/packaging/language/pip.py [hunk 3]
```diff
@@ -335,10 +335,11 @@ def _is_present(module, req, installed_pkgs, pkg_command):
     for pkg in installed_pkgs:
         if '==' in pkg:
             pkg_name, pkg_version = pkg.split('==')
+            pkg_name = Package.canonicalize_name(pkg_name)
```
Comment (webknjaz) [inline lib/ansible/modules/packaging/language/pip.py:353]: I'd probably move it to the initializer: https://github.com/ansible/ansible/issues/48095#issuecomment-436315372

File: lib/ansible/modules/packaging/language/pip.py [hunk 4]
```diff
@@ -523,6 +526,11 @@ def is_satisfied_by(self, version_to_test):
                 for op, ver in self._requirement.specs
             )
 
+    @staticmethod
+    def canonicalize_name(name):
+        # This is taken from PEP 503.
+        return Package._CANONICALIZE_RE.sub("-", name).lower()
```
Comment (webknjaz) [inline lib/ansible/modules/packaging/language/pip.py:548]: Which version of pip did this appear in?

**Comment by webknjaz** [comment]
Partially related issue, which will probably get fixed with this PR: #48095

File: lib/ansible/modules/packaging/language/pip.py (hunk 4 above)
Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py:548]: Tbh I have no idea. This is where I found it https://www.python.org/dev/peps/pep-0503/. I also found the exact code in pip, but since this a few months ago, I forgot :)

File: lib/ansible/modules/packaging/language/pip.py (hunk 2 above)
Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py]: I'm pretty sure I did it because it was reporting a long line. I must have changed this line before, so than it was long and I never returned it back. Anyway since this is where the pkg_list is initialized, this will have to be split into two lines, because PEP is failing with long line.

File: lib/ansible/modules/packaging/language/pip.py (hunk 3 above)
Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py:353]: Make sense. Will do.

**Comment by kustodian** [comment]
@webknjaz Yup issue #48095 is the same thing. So this PR will resolve #48095.

I made the suggested improvements, I guess we can get this merged once the tests pass.

Should we consider to add this into 2.7 release?

File: lib/ansible/modules/packaging/language/pip.py (hunk 3 above)
Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py:353]: Now I remember. It cannot be done in the initializer, because package name isn't split from the version there, so that means it would canonicalize the `name=version`, which isn't what we want. Anyway it's only needed here so it's fine that it stays.

File: lib/ansible/modules/packaging/language/pip.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/modules/packaging/language/pip.py:548]: Well, the reason I'm asking is that by using something which is present on only new envs you're adding an implicit dependency and we have to maintain backwards compatibility we committed to maintain.

**Comment by webknjaz** [comment]
@kustodian I think it is possible to get those in one of the upcoming bugfix releases, however I want to get a few more reviews from @HD650 and maybe @abadger to proceed. Also, I'll need to take time and review current changes more closely.

File: lib/ansible/modules/packaging/language/pip.py (hunk 4 above)
Comment (sivel) [reply lib/ansible/modules/packaging/language/pip.py:548]: I don't think we should be referencing a private attribute/variable here in any case.  We have no guarantee that it will exist in any future release either.

Comment (webknjaz) [reply lib/ansible/modules/packaging/language/pip.py:548]: Right, that too. Recently, pip moved all of "seemingly public" modules under explicit "_internal" package to make it clear that users ain't supposed to abuse those.

Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py:548]: I don't understand how is this not backward compatible even if older pip versions don't canonicalize package names. For example currently on pip it's not possible to create two different packages with these names:
* `Junit-XML`
* `junit-xml`
* `junit.xml`

This is a limitation of https://pypi.org/, so you will be able to have only one of those packages. I even did a fast test by counting the number of packages, and than canonicalizing their names and counting them again:
```
# Total number of packages
$ curl -s https://pypi.org/simple/ | grep -c href=
157503

# Number of packages after canonicalization
$ curl -s https://pypi.org/simple/ | grep href= | sed -r 's#[^"]+"/simple/([^/]+)/.*#\1#g' | tr '[:upper:]' '[:lower:]' | sort | wc -l
157503
```
As you can see it's the same number of packages.

That means even if we canonicalize names on older versions, nothing will change.

Comment (webknjaz) [reply lib/ansible/modules/packaging/language/pip.py:548]: I'm talking about `ImportError`.

**Comment by abadger** [comment]
I think since this is a constant we could set the constant based on whether
the constant appears in pip and if missing, fallback to the present value
of the constant.

Since it is a standard (the pep) i debated in my head whether we should
forgo the import from pip's private variables but in the end i think it has
value.  This module is attempting to give users an alternate front end to
pip, not just an alternate interface to the python packaging system which
happen to use pip as its backend.  Therefore if pip and the pep varied we
should follow what pip does.  Which we achieve by using the constant in pip
whenever we can find it.

There is a larger question of whether to use private interfaces of pip.  I
think the answer there is a balancing act between functionality and
fragility.  Depending on private interfaces is fragile.  So we should avoid
them when we can.  However, as i said above, we're really creating an
alternate interface to pip here.  So when pip does not give us public api
to create that interface and we are able to use the private interfaces
without tying the structure of our code to the private api (ie, calling pip
functions with some minimal structuring of the data before we call would be
easy to disentangle if we had to in the future.  Subclassing a pip data
structure and then overriding several of it's methods would not be) , we
are probably better off using the private Interface rather than reinventing
the code on our own.  Using the private interfaces has the fragility of pip
changing in a new version but it removes the fragility of the assumptions
our code makes about how we can mimic pip diverging from the algorithm that
pip itself makes.

If we were simply coding a generic "python package manager" here, instead
of an alternate front end to pip then this argument would not apply.

File: lib/ansible/modules/packaging/language/pip.py (hunk 4 above)
Comment (kustodian) [reply lib/ansible/modules/packaging/language/pip.py:548]: @webknjaz sorry, but I don't understand about what ImportError are you talking about? I didn't use any imports in this PR as far as I can see.

Comment (webknjaz) [reply lib/ansible/modules/packaging/language/pip.py:548]: Oh, I've got an impression that you're using pip's importables. In this case it's fine.

File: test/integration/targets/pip/tasks/pip.yml [hunk 5]
```diff
@@ -213,6 +213,27 @@
...
+    name: Junit-XML
+    virtualenv: "{{ output_dir }}/pipenv"
+    state: present
+  check_mode: True
+  register: diff_case_check_mode
+
+- name: make sure Junit-XML in check_mode doesn't report changed
+  assert:
+    that:
+      - "not (diff_case_check_mode is changed)"
```
Comment (webknjaz) [inline test/integration/targets/pip/tasks/pip.yml]: `is not`?

Comment (webknjaz) [reply test/integration/targets/pip/tasks/pip.yml]:
```suggestion
      - diff_case_check_mode is not changed
```

Comment (kustodian) [reply test/integration/targets/pip/tasks/pip.yml]: Fixed.

**Comment by kustodian** [comment]
I guess once the tests pass and when @webknjaz is finished reviewing, that we can merge this :)

File: lib/ansible/modules/packaging/language/pip.py [hunk 6]
```diff
@@ -499,6 +500,8 @@ class Package:
     test whether a package is already satisfied.
     """
 
+    _CANONICALIZE_RE = re.compile(r'[-_.]+')
```
Comment (webknjaz) [inline lib/ansible/modules/packaging/language/pip.py:503]: Shouldn't `.` and `-` be escaped properly?

Comment (webknjaz) [reply lib/ansible/modules/packaging/language/pip.py:503]: Okay, I've checked docs and it seems fine here.

**Comment by webknjaz** [comment]
@kustodian could you please add a news fragment to https://github.com/ansible/ansible/tree/devel/changelogs/fragments ?

**Review by webknjaz — APPROVED** [review]

**Comment by kustodian** [comment]
@webknjaz I added the fragment and squashed all commits into a single one. So this should be ready to be merged once the test pass.

**Review by HD650 — APPROVED** [review]

File: lib/ansible/modules/packaging/language/pip.py (hunk 6 above)
Comment (abadger) [reply lib/ansible/modules/packaging/language/pip.py:503]: Neither of those are special characters in python so they don't have to be escaped for python strings.  Inside of a character class regex characters like ```.``` don't have to be escaped because they don't have special meaning there. ```-``` doesn't have to be escaped in a character class as long as it is the first character in the character class.

**Comment by abadger** [comment]
Looks good to merge.  @kustodian or @webknjaz Can one of you take care of the backport?

**Comment by webknjaz** [comment]
@abadger here's the backport PR: #48437
